src/renderer/services/tab.ts

```diff
@@ -115,12 +127,99 @@ export const insertAtCursor = (tab: Tab, s: string): Tab => {
+    const newTab1 = { ...tab, content: newContent }
+    newTab1.diffs.done.push({
+        op: 'add',
+        at: newTab.cursor,
+        value: s.split('\n'),
+    })
+    newTab1.diffs.undone = []
+
+    return setCursor(newTab1, newCursor, false)
 }
 
 export const removeAtCursor = (tab: Tab): Tab => {
     const cursor = adjustedCursor(tab)
     const [newContent, newCursor] = removeSelection(tab.content, cursor)
 
-    return setCursor({ ...tab, content: newContent }, newCursor, false)
+    const newTab = { ...tab, content: newContent }
+
+    const cursorSelection = selection(cursor)
+    if (cursorSelection !== null) {
+        const normalizedCursorSelection = normalized(cursorSelection)
+        const value = subContent(tab.content, cursorSelection)
+
+        if (value !== null) {
+            newTab.diffs.done.push({
+                op: 'rm',
+                at: normalizedCursorSelection.start,
+                value,
+            })
+            newTab.diffs.undone = []
+        }
+    }
+
+    return setCursor(newTab, newCursor, false)
+}
+
+export const undoRedo = (tab: Tab, op: 'undo' | 'redo'): Tab => {
+    switch (op) {
+        case 'undo': {
+            if (tab.diffs.done.length === 0) {
+                return { ...tab }
+            }
+
+            const diff = tab.diffs.done[tab.diffs.done.length - 1]
+            const inverseDiff: ContentDiff = {
+                ...diff,
+                op: diff.op === 'add' ? 'rm' : 'add',
+            }
+
+            const [newContent, newCursor] = applyDiff(tab.content, inverseDiff)
+            const newDiffs = {
+                done: [...tab.diffs.done.slice(0, tab.diffs.done.length - 1)],
+                undone: [...tab.diffs.undone, inverseDiff],
+            }
+
+            return setCursor(
+                {
+                    ...tab,
+                    content: newContent,
+                    diffs: newDiffs,
+                },
+                newCursor,
+                false,
+            )
+        }
+
+        case 'redo': {
+            if (tab.diffs.undone.length === 0) {
+                return { ...tab }
+            }
+
+            const diff = tab.diffs.undone[tab.diffs.undone.length - 1]
+            const inverseDiff: ContentDiff = {
+                ...diff,
+                op: diff.op === 'add' ? 'rm' : 'add',
+            }
+
+            const [newContent, newCursor] = applyDiff(tab.content, inverseDiff)
+            const newDiffs = {
+                undone: [
+                    ...tab.diffs.undone.slice(0, tab.diffs.undone.length - 1),
+                ],
+                done: [...tab.diffs.done, inverseDiff],
+            }
+
+            return setCursor(
+                {
+                    ...tab,
+                    content: newContent,
+                    diffs: newDiffs,
+                },
+                newCursor,
+                false,
+            )
+        }
+    }
```

Comment:
Lots of duplicate code. Could have a function for moving the inverse diffs between the `done` and `undone` stacks.